lib/helpers.py

```diff
@@ -246,143 +340,248 @@ def _drop_duplicates_in_attributions(df, max_timedelta):
-            test_converters = test_revenue_df['user_id'].nunique()
-
-        if CONTROL in grouped_revenue.groups:
-            control_revenue_df = grouped_revenue.get_group(CONTROL)
-            control_revenue_micros = control_revenue_df['revenue_eur'].sum()
-            # control_conversions = control_revenue_df['partner_event'].count()
-            # as we filtered by revenue event and dropped the column we can just use
-            control_conversions = control_revenue_df['user_id'].count()
-            control_converters = control_revenue_df['user_id'].nunique()
-
-        # calculate KPIs
-        test_revenue = test_revenue_micros / 10 ** 6
-        control_revenue = control_revenue_micros / 10 ** 6
+        test_revenue_micros = test_users_df['revenue_eur'].sum()
+        test_conversions = test_users_df['conversion_count'].sum()
+        test_converters = (test_users_df['conversion_count'] > 0).sum()
+
+        control_revenue_micros = control_users_df['revenue_eur'].sum()
+        control_conversions = control_users_df['conversion_count'].sum()
+        control_converters = (control_users_df['conversion_count'] > 0).sum()
+
+        # samples for bootstrapping
+        control_group_conversions_sample = control_users_df['conversion_count']
+        control_group_revenue_sample = control_users_df['revenue_eur']
 
         ratio = float(test_group_size) / float(control_group_size)
+
+        # Converter KPIs
+        scaled_control_converters = float(control_converters) * ratio
+        no_treat_converters_lower_bound = scipy.stats.binom.ppf(
+            (1 - self.confidence_level) / 2,
+            control_group_size,
+            control_converters / control_group_size,
+        )
+        no_treat_converters_upper_bound = scipy.stats.binom.ppf(
+            ((1 - self.confidence_level) / 2) + self.confidence_level,
+            control_group_size,
+            control_converters / control_group_size,
+        )
+
+        incremental_converters_estimate = test_converters - control_converters * ratio
+
+        incremental_converters_lower_bound = test_converters - no_treat_converters_upper_bound * ratio
+        incremental_converters_upper_bound = test_converters - no_treat_converters_lower_bound * ratio
+
+        cost_per_incremental_converter_estimate = ad_spend / incremental_converters_estimate
+
+        cost_per_incremental_converter_lower_bound = ad_spend / incremental_converters_upper_bound
+        cost_per_incremental_converter_upper_bound = ad_spend / incremental_converters_lower_bound
+
+        if cost_per_incremental_converter_estimate < 0:
+            cost_per_incremental_converter_estimate = 'Negative incremental converters estimate'
+        if cost_per_incremental_converter_upper_bound < 0:
+            cost_per_incremental_converter_lower_bound = cost_per_incremental_converter_upper_bound = \
+                'CI contains negative value, cannot be interpreted'
```

Comment:
what does CI mean in this context? Is this know by users?

Reply:
In this context it's always `Confidence Interval(s)`
I can write it full though

Reply:
actually, since it's not the first or only abbreviation, let it be :)